src/main/java/cz/tstrecha/timetracker/service/impl/UserServiceImpl.java

```diff
@@ -163,6 +165,31 @@ public LoginResponseDTO changeContext(Long id, UserContext userContext) {
                 JwtAuthenticationFilter.AUTHORIZATION_HEADER_BEARER_PREFIX + refreshToken);
     }
 
+    @Override
+    @Transactional
+    public UserContext changeUserDetails(UserUpdateDTO userUpdateDTO, UserContext userContext) {
+        if (userUpdateDTO.getAccountType() == AccountType.PERSON) {
+            if (Strings.isEmpty(userUpdateDTO.getFirstName()) || Strings.isEmpty(userUpdateDTO.getLastName())) {
+                throw new UserInputException("Person has to have first name and last name filled in.",
+                        ErrorTypeCode.PERSON_FIRST_LAST_NAME_MISSING,
+                        "UserUpdateDTO");
+            }
+        } else if(userUpdateDTO.getAccountType() == AccountType.COMPANY){
+            if (Strings.isEmpty(userUpdateDTO.getCompanyName())){
+                throw new UserInputException("Company has to have company name filled in.",
+                        ErrorTypeCode.COMPANY_NAME_MISSING,
+                        "UserUpdateDTO");
+            }
+        }
+
+        var user = userRepository.findByEmail(userContext.getEmail()).orElseThrow();
+        userMapper.updateUser(userUpdateDTO, user);
+        userRepository.save(user);
+
+        return userMapper.toContext(user, userContext.getLoggedAs());
```

Comment:
When changing user, they can change their username. Frontend will decode using base64 context and show username from token. Therefore when changing username you have to provide a new token for user to ensure that correct and current information is being displayed to them